File: Queue/singlyLL.c
/*
** Author: Bryce Calhoun
** Description: Singly Linked List implemented with both a head and tail sentinel node.
*/

#include <stdlib.h>
#include <stdio.h>


struct Node{
    /*
    Description: Node structre defined with a void pointer and a 
                Node pointer to point to the next node in the list
    */
    void* value;
    struct Node* next;
};


struct singlyLL{
    /*
    Description: Singly linked list structre defined with an integer variable to track size 
                as well as a head sentinel and a tail sentinel
    */
    int size;
    struct Node* head_sentinel;
    struct Node* tail_sentinel;
};


struct singlyLL* create_list(){
    /*
    Description: Function to allocate memory for a singlyLL struct and 
                initialize its attributes. Function woll return a pointer 
                to the struct
    */
    struct singlyLL* list = malloc(sizeof(struct singlyLL));
    list->size = 0;

    list->head_sentinel = malloc(sizeof(struct Node));
    list->tail_sentinel = malloc(sizeof(struct Node));

    list->head_sentinel->next = NULL;
    list->tail_sentinel->next = NULL;

    return list;
}

int is_empty(struct singlyLL* list){
    /*
    Description: Function to return 1 (true) if a list is empty and 0 (false) otherwise
    */
    if(list->size == 0){
        return 1;
    }
    return 0;

}

void insert_front(struct singlyLL* list, void* value){
    /*
    Description: Function to take a void pointer and insert it in the front of a singlyLL 
                object
    */
    struct Node* newNode = malloc(sizeof(struct Node));
    newNode->value = value;
    int empty = is_empty(list);
    if(empty){
        newNode->next = NULL;
        list->head_sentinel->next = newNode;
        list->tail_sentinel->next = newNode;
    }
    else{
        newNode->next = list->head_sentinel->next;
        list->head_sentinel->next = newNode;
    }
    list->size++;
    return;
}

void* remove_front(struct singlyLL* list){
    /*
    Description: Function to remove a node from the front of a singlyLL
                and return the void pointer value
    */
    int empty = is_empty(list);
    if(empty){
        printf("EMPTY LIST");
        return NULL;
    }
    void* value = list->head_sentinel->next->value;

    struct Node* tempNode = list->head_sentinel->next;
    list->head_sentinel->next = list->head_sentinel->next->next;

    free(tempNode);

    list->size--;
    return value;
}

void insert_back(struct singlyLL* list, void* value){
    /*
    Description: Function to take a void pointer and insert it in the back of a singlyLL 
                object
    */
    struct Node* newNode = malloc(sizeof(struct Node));
    newNode->value = value;
    newNode->next = NULL;

    int empty = is_empty(list);
    if(empty){
        list->head_sentinel->next = newNode;
        list->tail_sentinel->next = newNode;
    }
    else{
        list->tail_sentinel->next->next = newNode;
        list->tail_sentinel->next = newNode;
    }

    list->size++;
    return;

}

void* get_front(struct singlyLL* list){
    int empty = is_empty(list);
    if(empty){
        printf("EMPTY LIST");
        return NULL;
    }
    
    void* value = list->head_sentinel->next->value;
    return value;

}

File: Linked_List/Singly/singlyLL.py
"""
#   Author: Bryce Calhoun
#   Description: A singly-linked list implemented with node structures.
"""

import random

class NodeException(Exception):

    pass

class Node:
    """
    DESCRIPTION: Node structure to represent each data point in the
                 linked list. Initialized with no value or next pointer.
    """
    def __init__(self, value=None, next=None):
        self._value = value
        self._next = next

    def get_value(self):
        if self._value is None:
            raise NodeException("NO VALUE IN NODE")
        return self._value

    def set_next(self, node):
        self._next = node

    def get_next(self):
        if self._next is None:
            raise NodeException("NO NEXT NODE")
        return self._next


class singlyLLException(Exception):

    pass


class singlyLL():
    """
    DESCRIPTION: Singly linked list initialized with head and tail sentinel nodes and
                 the size set to 0
    """
    def __init__(self):
        self._headSentinel = Node()
        self._tailSentinel = Node()
        self._size = 0

    def is_empty(self):
        """
        DESCRIPTION: function to return true if the list is empty, and false otherwise
        """
        if self._headSentinel.get_next() is None:
            return True
        return False

    def insert_front(self, value):
        """
        DESCRIPTION: Function to insert a value in the front of the list just
                     after the head sentinel
        """
        if self.is_empty():
            newNode = Node(value, None)
            self._headSentinel.set_next(newNode)
            self._tailSentinel.set_next(newNode)
            self._size += 1
            return

        newNode = Node(value, self._headSentinel.get_next())
        self._headSentinel.set_next(newNode)
        self._size += 1
        return

    def remove_front(self) -> object:
        """
        DESCRIPTION: Function to remove a value from the front of the list just
                     after the head sentinel
        """
        if self.is_empty():
            raise singlyLLException("EMPTY LIST")
        value = self._headSentinel.get_next().get_value()
        current = Node(None, self._headSentinel.get_next())
        current = current.get_next()
        self._headSentinel.set_next(current.get_next())
        self._size -= 1

        return value







